kong/plugins/moesif/moesif_gov.lua

```diff
@@ -0,0 +1,227 @@
+    local hash_key = ngx_md5(conf.application_id)
+    local user_id_entity
+    local company_id_entity
+    local request_headers = req_get_headers()
+
+    -- Set fetch_governance_rule to true (default), to be able to fetch governance rule
+    if fetch_governance_rules[hash_key] == nil then
+        fetch_governance_rules[hash_key] = true
+    end
+
+    if (governance_rules_etags[hash_key] == nil or (conf["rulesETag"] ~= governance_rules_etags[hash_key])) and fetch_governance_rules[hash_key] then
+        -- Set fetch_governance_rule to false, to avoid fetching governance rule
+        fetch_governance_rules[hash_key] = false
+
+        local gr_ok, gr_err = ngx_timer_at(0, get_governance_rules, hash_key, conf)
+        if not gr_ok then
+            -- Set fetch_governance_rule to true, to be able to fetch governance rule next time
+            fetch_governance_rules[hash_key] = true
+            if conf.debug then 
+                ngx_log(ngx_log_ERR, "[moesif] failed to get governance rules ", gr_err)
+            end
+        else
+            if conf.debug then 
+                ngx_log(ngx.DEBUG, "[moesif] successfully fetched the governance rules " , gr_ok)
+            end
+        end
+    end
+
+    -- Fetch the user details
+    if request_headers[conf.user_id_header] ~= nil then
+        user_id_entity = tostring(request_headers[conf.user_id_header])
+    elseif request_headers["x-consumer-custom-id"] ~= nil then
+        user_id_entity = tostring(request_headers["x-consumer-custom-id"])
+    elseif request_headers["x-consumer-username"] ~= nil then
+        user_id_entity = tostring(request_headers["x-consumer-username"])
+    elseif request_headers["x-consumer-id"] ~= nil then
+        user_id_entity = tostring(request_headers["x-consumer-id"])
+    else
+        user_id_entity = nil
+    end
+
+    -- Fetch the company details
+    if request_headers[conf.company_id_header] ~= nil then
+        company_id_entity = tostring(request_headers[conf.company_id_header])
+    else 
+        company_id_entity = nil
+    end
+
+    -- Set entity in conf to use downstream
+    conf["user_id_entity"] = user_id_entity
+    conf["company_id_entity"] = company_id_entity
+
+    if governance_rules_hashes[hash_key] ~= nil and type(governance_rules_hashes[hash_key]) == "table" and next(governance_rules_hashes[hash_key]) ~= nil then
+        -- Check if need to block request based on user governance rule
+        if user_id_entity ~= nil then 
+            local sc_user_req = block_request_based_on_entity_governance_rule(hash_key, conf, "user_rules", user_id_entity)
+            if sc_user_req == nil then 
+                if conf.debug then
+                    ngx_log(ngx.DEBUG, "[moesif] Skipped blocking request based on the user Id - " .. user_id_entity)
+                end
+            end
+        end
+        -- Check if need to block request based on company governance rule
+        if company_id_entity ~= nil then 
```

Comment:
check here that sc_user_req is  nil. No point in checking company rules again if its already blocked by  user rules

Reply:
If it is blocked by user rules, it won't check for company rules because the request will be short-circuited by `kong.response.exit` and no downstream logic would be executed in this phase.

Reply:
it doesn;t make sense even to call this fcn block_request_based_on_entity_governance_rule if block_request_based_on_entity_governance_rule already returned non nil. Lets add check here